stdlib/src/lib.rs

```diff
@@ -2,18 +2,21 @@
 
 extern crate alloc;
 
-use assembly::{
-    ast::Module, utils::Deserializable, Library, LibraryNamespace, LibraryPath, MaslLibrary,
-    Version,
-};
+use assembly::{library::CompiledLibrary, utils::Deserializable};
 
 // STANDARD LIBRARY
 // ================================================================================================
 
 /// TODO: add docs
-pub struct StdLibrary(MaslLibrary);
+pub struct StdLibrary(CompiledLibrary);
```

Comment:
I'm wondering if we actually need to define a new struct. An alternative could to have this crate expose a single method - something like:
```rust
pub fn get_library() -> CompiledLibrary
```

Reply:
I think it is nice to use wrapper structs that contain a `CompiledLibrary`, since that lets them be used in places where a `CompiledLibrary` is expected, without needing to extract the actual `CompiledLibrary` explicitly. In fact, I would suggest re-introducing some library traits, that looks like so:

```rust
/// Exposes `CompiledLibrary` functionality on any wrapper struct that impls AsRef,
/// but does not permit overriding the core `CompiledLibrary` APIs
///
/// Extra library-related functionality that allows customization, would be added here
pub trait Library: sealed::LibraryImpl {}

mod sealed {
    /// Ensures that only types that contain a `CompiledLibrary` can implement `Library`
    pub(super) trait LibraryImpl: AsRef<CompiledLibrary> {
        fn digest(&self) -> RpoDigest {
            self.as_ref().digest()
        }
        fn mast_forest(&self) -> &MastForest {
            self.as_ref().mast_forest()
       }
    }
    impl<T: AsRef<CompiledLibrary>> LibraryImpl for T {}
}

/// Marker trait for `Library` impls which are valid kernels
pub trait Kernel: Library + AsRef<KernelLibrary> {}


```

In virtually every place where we _use_ `CompiledLibrary` now, we actually expect something that impls `AsRef<CompiledLibrary>`, but I would argue it makes sense to have a dedicated trait for this, so that we can allow custom library types to customize certain behaviors.

It would also make it possible to instantiate a static `Library` impl, as opposed to constructing it every time someone calls `get_library()`.

Anyway, food for thought.

Reply:
One of the things I'm hoping to do is define the assembler in such a way that we construct it with a set of libraries upfront. This way, we don't need to deal with situations when libraries can be added dynamically during the assembly process (which, I think, in practice should never happen).

With `Library` being a struct, we could have the following constructor for the assembler:
```rust
impl Assembler {

    pub fn new(
        source_manager: Arc<dyn SourceManager>,
        kernel: KernelLibrary,
        libraries: impl IntoIterator<Item = Library>,
    ) -> Result<Self, Report> {
        ...
    }

}
```
We could have other, simpler constructors (e.g., `with_kernel()`) which would internally call this constructor - but the main idea is that once we've instantiated the assembler, we can only add modules and assembler programs/libraries etc.

If `Library` is a trait - this becomes a bit more complicated, I think.